crates/core/tedge/src/cli/c8y/upload.rs

```diff
@@ -0,0 +1,100 @@
+use crate::command::Command;
+use crate::log::MaybeFancy;
+use anyhow::Error;
+use c8y_api::http_proxy::C8yEndPoint;
+use c8y_api::json_c8y::C8yCreateEvent;
+use c8y_api::json_c8y::C8yEventResponse;
+use c8y_api::json_c8y::C8yManagedObject;
+use c8y_api::json_c8y::InternalIdResponse;
+use c8y_api::OffsetDateTime;
+use reqwest::blocking;
+use reqwest::blocking::multipart;
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+/// Upload a file to Cumulocity
+pub struct C8yUpload {
+    /// Device identifier
+    pub device_id: String,
+
+    /// Cumulocity endpoint
+    pub c8y: C8yEndPoint,
+
+    /// Type of the event.
+    pub event_type: String,
+
+    /// Text description of the event.
+    pub text: String,
+
+    /// JSON fragment attached to the event
+    pub json: HashMap<String, serde_json::Value>,
+
+    /// Path to the uploaded file
+    pub file: PathBuf,
+
+    /// MIME type of the file content. Defaults to `application/octet-stream`
+    pub mime_type: String,
+}
+
+impl Command for C8yUpload {
+    fn description(&self) -> String {
+        "upload a file to Cumulocity".to_string()
+    }
+
+    fn execute(&self) -> Result<(), MaybeFancy<Error>> {
+        let internal_id = self.get_internal_id()?;
+        let event_id = self.create_event(internal_id)?;
+        self.upload_file(&event_id)?;
+
+        println!("{event_id}");
+        Ok(())
+    }
+}
+
+impl C8yUpload {
+    pub fn get_internal_id(&self) -> Result<String, Error> {
+        let url_get_id: String = self.c8y.proxy_url_for_internal_id(&self.device_id);
+        let http_result = blocking::Client::new().get(url_get_id).send()?;
```

Comment:
As stressed by clippy, I have to use `certificate::CloudRootCerts` type instead to take root_cert_path configurations into account, i.e. to make tedge c8y upload works when the c8y proxy is configured with mTLS.